numba/typing/builtins.py

```diff
@@ -94,16 +94,7 @@ def generic(self, args, kws):
         assert not kws
         [obj] = args
         if isinstance(obj, types.IterableType):
-            # Raise this here to provide a very specific message about this
-            # common issue, delaying the error until later leads to something
-            # less specific being noted as the problem (e.g. no support for
-            # getiter on array(<>, 2, 'C')).
-            if isinstance(obj, types.Array) and obj.ndim > 1:
```

Comment:
This branch is still needed for the case where `obj.ndim == 0`.

It can happen with:

```python
@njit
def foo(a):
    return list(a)

foo(np.array(0))
```
And this case needs to be tested.

Reply:
I don't understand what you mean. When called on that input, this `if` is not executed, because `obj.ndim > 1` is false. Both before and after this patch, that code path runs `return signature(obj.iterator_type, obj)`.

Reply:
Can I leave fixing 0d arrays to a separate patch? The fix is unrelated, and I don't know how best to write the test.

The fix for 0d arrays is:
```diff
--- a/numba/types/iterators.py
+++ b/numba/types/iterators.py
@@ -99,7 +99,9 @@ class ArrayIterator(SimpleIteratorType):
         self.array_type = array_type
         name = "iter(%s)" % (self.array_type,)
         nd = array_type.ndim
-        if nd == 0 or nd == 1:
+        if nd == 0:
+            raise errors.TypingError("iteration over a 0-d array")
+        elif nd == 1:
             yield_type = array_type.dtype
         else:
             yield_type = array_type.copy(ndim=array_type.ndim - 1)
```

Reply:
Error before this PR:
```
LoweringError: Failed in nopython mode pipeline (step: nopython mode backend)
Failed in nopython mode pipeline (step: nopython mode backend)
Failed in nopython mode pipeline (step: nopython mode backend)
iterating over 0D array

File "..\..\..\AppData\Roaming\Python\Python38\site-packages\numba\targets\listobj.py", line 902:
    def list_extend(lst, iterable):
        <source elided>
        meth = lst.append
        for v in iterable:
```
Error with this PR:
```
LoweringError: Failed in nopython mode pipeline (step: nopython mode backend)
Failed in nopython mode pipeline (step: nopython mode backend)
Failed in nopython mode pipeline (step: nopython mode backend)
iterating over 0D array

File "AppData\Roaming\Python\Python38\site-packages\numba\targets\listobj.py", line 902:
    def list_extend(lst, iterable):
        <source elided>
        meth = lst.append
        for v in iterable:
```

Reply:
I'm seeing this error:

```pytb
numba.errors.LoweringError: Failed in nopython mode pipeline (step: nopython mode backend)
Failed in nopython mode pipeline (step: nopython mode backend)
Failed in nopython mode pipeline (step: nopython mode backend)
Can't index at [0] in [0 x i64]

File "numba/targets/listobj.py", line 902:
    def list_extend(lst, iterable):
        <source elided>
        meth = lst.append
        for v in iterable:
        ^
```

which is complaining from the llvmlite layer.

Reply:
I am testing it by running:

```python
from numba import njit
import numpy as np

@njit
def foo(a):
    return list(a)

foo(np.array(0))
```

Reply:
Looks like I have some old numba installs clobbering my branch installation

Reply:
How's this?

```
In [2]: @numba.njit
   ...: def foo(a):
   ...:     return list(a)
   ...:
   ...: foo(np.array(0))
---------------------------------------------------------------------------
TypingError                               Traceback (most recent call last)
<ipython-input-2-8fd610bfe89b> in <module>
      3     return list(a)
      4
----> 5 foo(np.array(0))

c:\users\wiese\repos\ga\numba\numba\dispatcher.py in _compile_for_args(self, *args, **kws)
    399                 e.patch_message(msg)
    400
--> 401             error_rewrite(e, 'typing')
    402         except errors.UnsupportedError as e:
    403             # Something unsupported is present in the user code, add help info

c:\users\wiese\repos\ga\numba\numba\dispatcher.py in error_rewrite(e, issue_type)
    342                 raise e
    343             else:
--> 344                 reraise(type(e), e, None)
    345
    346         argtypes = []

c:\users\wiese\repos\ga\numba\numba\six.py in reraise(tp, value, tb)
    666             value = tp()
    667         if value.__traceback__ is not tb:
--> 668             raise value.with_traceback(tb)
    669         raise value
    670

TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Invalid use of Function(<class 'list'>) with argument(s) of type(s): (array(int32, 0d, C))
 * parameterized
In definition 0:
    TypingError: iteration over a 0-d array
    raised from c:\users\wiese\repos\ga\numba\numba\types\iterators.py:104
In definition 1:
    TypingError: iteration over a 0-d array
    raised from c:\users\wiese\repos\ga\numba\numba\types\iterators.py:104
This error is usually caused by passing an argument of a type that is unsupported by the named function.
[1] During: resolving callee type: Function(<class 'list'>)
[2] During: typing of call at <ipython-input-2-8fd610bfe89b> (3)


File "<ipython-input-2-8fd610bfe89b>", line 3:
def foo(a):
    return list(a)
    ^
```